milestones/data.py

```diff
@@ -0,0 +1,373 @@
+
+2) Calculating derivative information from existing state:
+* Algorithms and data manipulations
+* Aggregations
+* Annotations
+* Alternative data representations
+
+Accepts and returns standard Python data structures (dicts, arrays of dicts)
+for easy consumption and manipulation by callers -- the queryset stops here!
+
+Note the terminology difference at this layer vs. API -- create/fetch/update/delete/
+
+When the time comes for remote resources, import the module like so:
+if getattr(settings, 'TEST_MODE', False) or os.getenv('TRAVIS_MODE', False):
+    import milestones.tests.mocks.resources as remote
+else:
+    import milestones.resources as remote
+"""
+from . import exceptions
+from . import models as internal
+from . import serializers
+
+
+# PRIVATE/INTERNAL METHODS
+def _get_milestone_relationship_type(relationship):
+    """
+    Retrieves milestone relationship type object from backend
+    """
+    try:
+        return internal.MilestoneRelationshipType.objects.get(
+            name=relationship,
+            active=True
+        )
+    except internal.MilestoneRelationshipType.DoesNotExist:
+        if relationship in ['requires', 'fulfills']:
+            return internal.MilestoneRelationshipType.objects.create(
+                name=relationship,
+                active=True
+            )
+        else:
+            raise exceptions.InvalidMilestoneRelationshipTypeException()
+
+
+# PUBLIC METHODS
+def create_milestone(milestone):
+    """
+    Inserts a new milestone into app/local state
+    Returns a dictionary representation of the object
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    milestone, __ = internal.Milestone.objects.get_or_create(  # pylint: disable=invalid-name
+        namespace=milestone_obj.namespace,
+        name=milestone_obj.name,
+        active=True,
+        defaults={
+            'description': milestone_obj.description,
+        }
+    )
+    return serializers.serialize_milestone(milestone)
+
+
+def update_milestone(milestone):
+    """
+    Updates an existing milestone in app/local state
+    Returns a dictionary representation of the object
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    try:
+        milestone = internal.Milestone.objects.get(id=milestone_obj.id)
+        milestone.name = milestone_obj.name
+        milestone.namespace = milestone_obj.namespace
+        milestone.description = milestone_obj.description
+        milestone.active = milestone_obj.active
+    except internal.Milestone.DoesNotExist:
+        raise exceptions.InvalidMilestoneException()
+    return serializers.serialize_milestone(milestone)
+
+
+def delete_milestone(milestone):
+    """
+    Deletes an existing milestone from app/local state
+    No return currently defined for this operation
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    _delete_milestone(milestone_obj)
```

Comment:
I agree that it would be nicer to go with the soft deletes (ie, set active=False instead of actually removing records).  As you've noticed all of the filtering is in place to treat inactive records as deleted on the read-side, so I think it's just a matter of making a call one way or the other -- @marcotuts, what do you think?  I'll also catch up with @ormsbee about this tomorrow.